File: macos-app/Core/Sources/MiniVPNCore/ViewModels/ConnectionViewModel.swift
import Foundation
import Combine

@MainActor
public final class ConnectionViewModel: ObservableObject {
    @Published public private(set) var state: ConnectionState = .disconnected
    @Published public private(set) var upBps: Int = 0
    @Published public private(set) var downBps: Int = 0
    @Published public private(set) var upBytes: Int = 0
    @Published public private(set) var downBytes: Int = 0
    @Published public private(set) var logs: [LogLine] = []

    private let control: ControlService
    private var pump: Task<Void, Never>?

    public init(control: ControlService) {
        self.control = control
    }

    /// Begin draining the control event stream into published state.
    public func start() {
        pump?.cancel()
        let stream = control.events()
        pump = Task { [weak self] in
            for await event in stream {
                guard let self else { return }
                await self.apply(event)
            }
        }
    }

    private func apply(_ event: ControlEvent) {
        switch event {
        case .state(let s): state = s
        case .stats(let st):
            upBps = st.upBps; downBps = st.downBps
            upBytes = st.upBytes; downBytes = st.downBytes
        case .log(let line):
            append(line)
        case .error(let detail):
            state = .error
            append(LogLine(level: .error, message: detail, ts: Date()))
        }
    }

    private func append(_ line: LogLine) {
        logs.append(line)
        if logs.count > 500 { logs.removeFirst(logs.count - 500) }
    }

    /// Command helpers absorb a `send` failure into `.error` + a log so the call
    /// sites (views) stay non-throwing — this is what keeps the mock→real swap
    /// from rippling into the UI.
    public func connect(nodeId: String) async { await send(.connect(nodeId: nodeId)) }
    public func auto() async { await send(.auto) }
    public func disconnect() async { await send(.disconnect) }

    private func send(_ command: ControlCommand) async {
        do {
            try await control.send(command)
        } catch {
            state = .error
            append(LogLine(level: .error, message: "\(error)", ts: Date()))
        }
    }

    public var isConnected: Bool { state == .connected }
}

File: android-app/app/src/main/java/com/minivpn/app/control/ConnectionViewModel.kt
package com.minivpn.app.control

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import uniffi.minivpn_core.ConnectionState
import uniffi.minivpn_core.ControlCommand
import uniffi.minivpn_core.ControlEvent
import uniffi.minivpn_core.ControlService
import uniffi.minivpn_core.EventObserver

/** UI-facing snapshot of the rust-core ① ControlService event stream. */
data class ConnectionUiState(
    val state: ConnectionState = ConnectionState.DISCONNECTED,
    val upBps: Long = 0,
    val downBps: Long = 0,
    val upBytes: Long = 0,
    val downBytes: Long = 0,
    val lastLog: String = "",
)

/**
 * Thin Kotlin VM over the rust-core `ControlService` (per ADR-C "中核": the
 * state machine lives in Rust; this wraps its event stream into a [StateFlow]
 * that Compose collects). `send(Connect/Disconnect)` drives it over FFI.
 *
 * THREADING (FINDINGS §8): `EventObserver.onEvent` is invoked on a Rust thread
 * (the ticker thread), so we hop to `Dispatchers.Main` before mutating the
 * StateFlow the UI observes — the Kotlin analogue of the Swift MainActor hop.
 */
class ConnectionViewModel : ViewModel() {
    private val _ui = MutableStateFlow(ConnectionUiState())
    val ui: StateFlow<ConnectionUiState> = _ui.asStateFlow()

    private val observer = object : EventObserver {
        override fun onEvent(event: ControlEvent) {
            // Callback arrives off a Rust thread → hop to Main, then apply.
            viewModelScope.launch(Dispatchers.Main) { apply(event) }
        }
    }

    private val service = ControlService(observer, liveTicker = true)

    private fun apply(event: ControlEvent) {
        // Tag "MiniVPN" so a headless `adb logcat` run can verify the live
        // rust-core → FFI → Main path (the events originate in Rust).
        Log.i("MiniVPN", "[${Thread.currentThread().name}] $event")
        _ui.value = when (event) {
            is ControlEvent.State -> _ui.value.copy(state = event.state)
            is ControlEvent.Stats -> _ui.value.copy(
                upBps = event.upBps,
                downBps = event.downBps,
                upBytes = event.upBytes,
                downBytes = event.downBytes,
            )
            is ControlEvent.Log -> _ui.value.copy(lastLog = "${event.level}: ${event.message}")
            is ControlEvent.Error -> _ui.value.copy(lastLog = "error: ${event.detail}")
        }
    }

    fun connect() = service.send(ControlCommand.Connect("placeholder-node"))

    // rust-core `stop()` joins the ticker thread (up to ~1s mid-sleep), so keep
    // it off the main thread. Resulting Disconnected/Log events still flow back
    // through the observer's Main hop.
    fun disconnect() {
        viewModelScope.launch(Dispatchers.IO) { service.send(ControlCommand.Disconnect) }
    }

    override fun onCleared() {
        super.onCleared()
        // Retire the Rust ticker thread, then free the FFI handle.
        service.send(ControlCommand.Disconnect)
        service.close()
    }
}
